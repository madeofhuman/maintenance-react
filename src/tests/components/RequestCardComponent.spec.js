import React from 'react';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { BrowserRouter, Switch } from 'react-router-dom';
import RequestCard from '../../app/components/RequestCardComponent';

Enzyme.configure({ adapter: new Adapter() });

const requestInReview = {
  id: 29,
  type: 'repair',
  item: 'Yomis Head',
  model: 'N/A',
  detail: 'It spoiled.',
  status: 'in-review',
  owner: 'simonandal@gmail.com',
  created_at: '2018-10-05T09:51:47.773Z',
  updated_at: null,
};

const approvedRequest = {
  id: 29,
  type: 'repair',
  item: 'Yomis Head',
  model: 'N/A',
  detail: 'It spoiled.',
  status: 'pending',
  owner: 'simonandal@gmail.com',
  created_at: '2018-10-05T09:51:47.773Z',
  updated_at: null,
};

const resolvedRequest = {
  id: 29,
  type: 'repair',
  item: 'Yomis Head',
  model: 'N/A',
  detail: 'It spoiled.',
  status: 'resolved',
  owner: 'simonandal@gmail.com',
  created_at: '2018-10-05T09:51:47.773Z',
  updated_at: null,
};

const disapprovedRequest = {
  id: 29,
  type: 'repair',
  item: 'Yomis Head',
  model: 'N/A',
  detail: 'It spoiled.',
  status: 'disapproved',
  owner: 'simonandal@gmail.com',
  created_at: '2018-10-05T09:51:47.773Z',
  updated_at: null,
};

const requestActions = {
  approveRequest: () => jest.fn(),
  disapproveRequest: () => jest.fn(),
  resolveRequest: () => jest.fn(),
  deleteRequest: () => jest.fn(),
};

describe('RequestCard Component', () => {
  test('renders', () => {
    const wrapper = shallow(
      <RequestCard
        request={requestInReview}
        role="user"
        requestActions={requestActions}
      />,
    );
    expect(wrapper.exists()).toBe(true);
  });
  test('displays an Approve and a Disapprove button when a request is under review and the user is an admin', () => {
    const wrapper = mount(
      <BrowserRouter>
        <Switch>
          <RequestCard
            request={requestInReview}
            role="admin"
            requestActions={requestActions}
          />
        </Switch>
      </BrowserRouter>,
    );
    expect(wrapper.exists()).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('#approve-btn')).toHaveLength(2);
    expect(wrapper.find('#disapprove-btn')).toHaveLength(2);
  });
  test('displays a resolve button when a request has been approved and the user is an admin', () => {
    const wrapper = mount(
      <BrowserRouter>
        <Switch>
          <RequestCard
            request={approvedRequest}
            role="admin"
            requestActions={requestActions}
          />
        </Switch>
      </BrowserRouter>,
    );
    expect(wrapper.exists()).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('#resolve-btn')).toHaveLength(2);
  });
  test('displays a Request Completed messaged when a request has been resolved and the user is an admin', () => {
    const wrapper = mount(
      <BrowserRouter>
        <Switch>
          <RequestCard
            request={resolvedRequest}
            role="admin"
            requestActions={requestActions}
          />
        </Switch>
      </BrowserRouter>,
    );
    expect(wrapper.exists()).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('input').props().value).toEqual('Request Completed');
  });
  test('displays a disapproved button when the request has been disapproved and user is an admin', () => {
    const wrapper = mount(
      <BrowserRouter>
        <Switch>
          <RequestCard
            request={disapprovedRequest}
            role="admin"
            requestActions={requestActions}
          />
        </Switch>
      </BrowserRouter>,
    );
    expect(wrapper.exists()).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('input').props().value).toEqual('Request Disapproved');
  });
  test('displays an edit and a delete button when a request is in review and user is a user', () => {
    const wrapper = mount(
      <BrowserRouter>
        <Switch>
          <RequestCard
            request={requestInReview}
            role="user"
            requestActions={requestActions}
          />
        </Switch>
      </BrowserRouter>,
    );
    expect(wrapper.exists()).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('#edit-btn')).toHaveLength(2);
    expect(wrapper.find('#delete-btn')).toHaveLength(2);
  });
  test('displays no action buttons when a request has been approved, disapproved or resolved and the user is a user', () => {
    const wrapper = mount(
      <BrowserRouter>
        <Switch>
          <RequestCard
            request={approvedRequest}
            role="user"
            requestActions={requestActions}
          />
        </Switch>
      </BrowserRouter>,
    );
    expect(wrapper.exists()).toBe(true);
    expect(toJson(wrapper)).toMatchSnapshot();
    expect(wrapper.find('input')).toHaveLength(0);
  });
});
