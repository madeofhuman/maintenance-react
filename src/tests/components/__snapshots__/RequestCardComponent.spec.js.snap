// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestCard Component displays a Request Completed messaged when a request has been resolved and the user is an admin 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Switch>
      <RequestCard
        computedMatch={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        request={
          Object {
            "created_at": "2018-10-05T09:51:47.773Z",
            "detail": "It spoiled.",
            "id": 29,
            "item": "Yomis Head",
            "model": "N/A",
            "owner": "simonandal@gmail.com",
            "status": "resolved",
            "type": "repair",
            "updated_at": null,
          }
        }
        requestActions={
          Object {
            "approveRequest": [Function],
            "deleteRequest": [Function],
            "disapproveRequest": [Function],
            "resolveRequest": [Function],
          }
        }
        role="admin"
      >
        <div
          className="card white"
        >
          <div
            className="row"
          >
            <h2
              className="left orange"
              id="title"
            >
              Yomis Head
              , 
              repair
            </h2>
            <p
              className="right"
              id="date"
            >
              Fri Oct 05 2018
            </p>
          </div>
          <small
            className="left white small"
            id="model"
          >
            N/A
             -
          </small>
          <em
            id="status"
          >
             
            resolved
          </em>
          <br />
          <div
            className="row"
          >
            <div
              className="left"
              id="owner"
            >
              by simonandal@gmail.com
            </div>
          </div>
          <br />
          <div
            className="row"
          >
            <em
              id="detail"
            >
              It spoiled.
            </em>
          </div>
          <br />
          <br />
          <br />
          <div
            className="row center"
            id="request-btns"
          >
            <Input
              className="button left green-bg white"
              id="resolve-btn"
              isRequired=""
              type="button"
              value="Request Completed"
            >
              <label />
              <input
                className="button left green-bg white"
                id="resolve-btn"
                required=""
                type="button"
                value="Request Completed"
              />
            </Input>
          </div>
        </div>
      </RequestCard>
    </Switch>
  </Router>
</BrowserRouter>
`;

exports[`RequestCard Component displays a disapproved button when the request has been disapproved and user is an admin 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Switch>
      <RequestCard
        computedMatch={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        request={
          Object {
            "created_at": "2018-10-05T09:51:47.773Z",
            "detail": "It spoiled.",
            "id": 29,
            "item": "Yomis Head",
            "model": "N/A",
            "owner": "simonandal@gmail.com",
            "status": "disapproved",
            "type": "repair",
            "updated_at": null,
          }
        }
        requestActions={
          Object {
            "approveRequest": [Function],
            "deleteRequest": [Function],
            "disapproveRequest": [Function],
            "resolveRequest": [Function],
          }
        }
        role="admin"
      >
        <div
          className="card white"
        >
          <div
            className="row"
          >
            <h2
              className="left orange"
              id="title"
            >
              Yomis Head
              , 
              repair
            </h2>
            <p
              className="right"
              id="date"
            >
              Fri Oct 05 2018
            </p>
          </div>
          <small
            className="left white small"
            id="model"
          >
            N/A
             -
          </small>
          <em
            id="status"
          >
             
            disapproved
          </em>
          <br />
          <div
            className="row"
          >
            <div
              className="left"
              id="owner"
            >
              by simonandal@gmail.com
            </div>
          </div>
          <br />
          <div
            className="row"
          >
            <em
              id="detail"
            >
              It spoiled.
            </em>
          </div>
          <br />
          <br />
          <br />
          <div
            className="row center"
            id="request-btns"
          >
            <Input
              className="button left green-bg white"
              id="resolve-btn"
              isRequired=""
              type="button"
              value="Request Disapproved"
            >
              <label />
              <input
                className="button left green-bg white"
                id="resolve-btn"
                required=""
                type="button"
                value="Request Disapproved"
              />
            </Input>
          </div>
        </div>
      </RequestCard>
    </Switch>
  </Router>
</BrowserRouter>
`;

exports[`RequestCard Component displays a resolve button when a request has been approved and the user is an admin 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Switch>
      <RequestCard
        computedMatch={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        request={
          Object {
            "created_at": "2018-10-05T09:51:47.773Z",
            "detail": "It spoiled.",
            "id": 29,
            "item": "Yomis Head",
            "model": "N/A",
            "owner": "simonandal@gmail.com",
            "status": "pending",
            "type": "repair",
            "updated_at": null,
          }
        }
        requestActions={
          Object {
            "approveRequest": [Function],
            "deleteRequest": [Function],
            "disapproveRequest": [Function],
            "resolveRequest": [Function],
          }
        }
        role="admin"
      >
        <div
          className="card white"
        >
          <div
            className="row"
          >
            <h2
              className="left orange"
              id="title"
            >
              Yomis Head
              , 
              repair
            </h2>
            <p
              className="right"
              id="date"
            >
              Fri Oct 05 2018
            </p>
          </div>
          <small
            className="left white small"
            id="model"
          >
            N/A
             -
          </small>
          <em
            id="status"
          >
             
            pending
          </em>
          <br />
          <div
            className="row"
          >
            <div
              className="left"
              id="owner"
            >
              by simonandal@gmail.com
            </div>
          </div>
          <br />
          <div
            className="row"
          >
            <em
              id="detail"
            >
              It spoiled.
            </em>
          </div>
          <br />
          <br />
          <br />
          <div
            className="row center"
            id="request-btns"
          >
            <Input
              className="button left green-bg white"
              handleClick={[Function]}
              id="resolve-btn"
              isRequired=""
              type="button"
              value="Resolve"
            >
              <label />
              <input
                className="button left green-bg white"
                id="resolve-btn"
                onClick={[Function]}
                required=""
                type="button"
                value="Resolve"
              />
            </Input>
          </div>
        </div>
      </RequestCard>
    </Switch>
  </Router>
</BrowserRouter>
`;

exports[`RequestCard Component displays an Approve and a Disapprove button when a request is under review and the user is an admin 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Switch>
      <RequestCard
        computedMatch={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        request={
          Object {
            "created_at": "2018-10-05T09:51:47.773Z",
            "detail": "It spoiled.",
            "id": 29,
            "item": "Yomis Head",
            "model": "N/A",
            "owner": "simonandal@gmail.com",
            "status": "in-review",
            "type": "repair",
            "updated_at": null,
          }
        }
        requestActions={
          Object {
            "approveRequest": [Function],
            "deleteRequest": [Function],
            "disapproveRequest": [Function],
            "resolveRequest": [Function],
          }
        }
        role="admin"
      >
        <div
          className="card white"
        >
          <div
            className="row"
          >
            <h2
              className="left orange"
              id="title"
            >
              Yomis Head
              , 
              repair
            </h2>
            <p
              className="right"
              id="date"
            >
              Fri Oct 05 2018
            </p>
          </div>
          <small
            className="left white small"
            id="model"
          >
            N/A
             -
          </small>
          <em
            id="status"
          >
             
            in-review
          </em>
          <br />
          <div
            className="row"
          >
            <div
              className="left"
              id="owner"
            >
              by simonandal@gmail.com
            </div>
          </div>
          <br />
          <div
            className="row"
          >
            <em
              id="detail"
            >
              It spoiled.
            </em>
          </div>
          <br />
          <br />
          <br />
          <div
            className="row center"
            id="request-btns"
          >
            <br />
            <br />
            <Input
              className="button left green-bg white"
              handleClick={[Function]}
              id="approve-btn"
              isRequired=""
              type="button"
              value="Approve"
            >
              <label />
              <input
                className="button left green-bg white"
                id="approve-btn"
                onClick={[Function]}
                required=""
                type="button"
                value="Approve"
              />
            </Input>
            <br />
            <Input
              className="button left green-bg white"
              handleClick={[Function]}
              id="disapprove-btn"
              isRequired=""
              type="button"
              value="Disapprove"
            >
              <label />
              <input
                className="button left green-bg white"
                id="disapprove-btn"
                onClick={[Function]}
                required=""
                type="button"
                value="Disapprove"
              />
            </Input>
            <br />
            <br />
          </div>
        </div>
      </RequestCard>
    </Switch>
  </Router>
</BrowserRouter>
`;

exports[`RequestCard Component displays an edit and a delete button when a request is in review and user is a user 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Switch>
      <RequestCard
        computedMatch={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        request={
          Object {
            "created_at": "2018-10-05T09:51:47.773Z",
            "detail": "It spoiled.",
            "id": 29,
            "item": "Yomis Head",
            "model": "N/A",
            "owner": "simonandal@gmail.com",
            "status": "in-review",
            "type": "repair",
            "updated_at": null,
          }
        }
        requestActions={
          Object {
            "approveRequest": [Function],
            "deleteRequest": [Function],
            "disapproveRequest": [Function],
            "resolveRequest": [Function],
          }
        }
        role="user"
      >
        <div
          className="card white"
        >
          <div
            className="row"
          >
            <h2
              className="left orange"
              id="title"
            >
              Yomis Head
              , 
              repair
            </h2>
            <p
              className="right"
              id="date"
            >
              Fri Oct 05 2018
            </p>
          </div>
          <small
            className="left white small"
            id="model"
          >
            N/A
             -
          </small>
          <em
            id="status"
          >
             
            in-review
          </em>
          <br />
          <div
            className="row"
          >
            <div
              className="left"
              id="owner"
            >
              by simonandal@gmail.com
            </div>
          </div>
          <br />
          <div
            className="row"
          >
            <em
              id="detail"
            >
              It spoiled.
            </em>
          </div>
          <br />
          <br />
          <br />
          <div
            className="row center"
            id="request-btns"
          >
            <br />
            <br />
            <Input
              className="button left green-bg white"
              id="edit-btn"
              isRequired=""
              name="edit-request"
              type="button"
              value="       Edit       "
            >
              <label
                htmlFor="edit-request"
              />
              <input
                className="button left green-bg white"
                id="edit-btn"
                name="edit-request"
                required=""
                type="button"
                value="       Edit       "
              />
            </Input>
            <br />
            <Input
              className="button left green-bg white"
              handleClick={[Function]}
              id="delete-btn"
              isRequired=""
              type="button"
              value="    Delete    "
            >
              <label />
              <input
                className="button left green-bg white"
                id="delete-btn"
                onClick={[Function]}
                required=""
                type="button"
                value="    Delete    "
              />
            </Input>
            <br />
            <br />
          </div>
        </div>
      </RequestCard>
    </Switch>
  </Router>
</BrowserRouter>
`;

exports[`RequestCard Component displays no action buttons when a request has been approved, disapproved or resolved and the user is a user 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Switch>
      <RequestCard
        computedMatch={
          Object {
            "isExact": true,
            "params": Object {},
            "path": "/",
            "url": "/",
          }
        }
        location={
          Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          }
        }
        request={
          Object {
            "created_at": "2018-10-05T09:51:47.773Z",
            "detail": "It spoiled.",
            "id": 29,
            "item": "Yomis Head",
            "model": "N/A",
            "owner": "simonandal@gmail.com",
            "status": "pending",
            "type": "repair",
            "updated_at": null,
          }
        }
        requestActions={
          Object {
            "approveRequest": [Function],
            "deleteRequest": [Function],
            "disapproveRequest": [Function],
            "resolveRequest": [Function],
          }
        }
        role="user"
      >
        <div
          className="card white"
        >
          <div
            className="row"
          >
            <h2
              className="left orange"
              id="title"
            >
              Yomis Head
              , 
              repair
            </h2>
            <p
              className="right"
              id="date"
            >
              Fri Oct 05 2018
            </p>
          </div>
          <small
            className="left white small"
            id="model"
          >
            N/A
             -
          </small>
          <em
            id="status"
          >
             
            pending
          </em>
          <br />
          <div
            className="row"
          >
            <div
              className="left"
              id="owner"
            >
              by simonandal@gmail.com
            </div>
          </div>
          <br />
          <div
            className="row"
          >
            <em
              id="detail"
            >
              It spoiled.
            </em>
          </div>
          <br />
          <br />
          <br />
          <div
            className="row center"
            id="request-btns"
          />
        </div>
      </RequestCard>
    </Switch>
  </Router>
</BrowserRouter>
`;
